Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp

```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
     ASSERT(userContentFilter);
     return *userContentFilter->contentRuleList;
 }
+
+/**
+ * WebKitUserScriptMessage: (ref-func webkit_user_script_message_ref) (unref-func webkit_user_script_message_unref)
+ *
+ * A helper class that sends user provided script to the signal handler.
+ *
+ * Since: 2.38
+ */
+struct _WebKitUserScriptMessage {
+    _WebKitUserScriptMessage(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
+        : completionHandler(WTFMove(completionHandler))
+        , referenceCount(1)
+    {
+        JSCContext* jscContext = SharedJavascriptContext::singleton().getOrCreateContext();
+        JSValueRef jsValueRef = serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr);
+        message = jscValueCreate(jscContext, jsValueRef);
+    }
+
+    GRefPtr<JSCValue> message;
+    GRefPtr<JSCValue> replyValue;
+    GUniquePtr<GError> replyError;
```

Comment:
I think in the end we do not really need to save these. We can invoke the `completionHandler` directly inside `webkit_user_script_message_reply()` and `webkit_user_script_message_error()`.

Reply:
```suggestion
```